docs: Add quickstart guide. (#6782)
cvc5/cvc5: docs/api/cpp/quickstart.rst

Quickstart Guide
================

> First, create a cvc5 :cpp:class:`Solver <cvc5::api::Solver>` instance:

Reviewer: Probably not for this PR, but I wonder what do you think about adding details on how to build this example, that is, starting one step before.

Author: Hm. This is the quickstart guide for the API, not sure if we should have build instructions there. I'm thinking this should probably rather be mentioned in https://cvc5.github.io/docs/examples/examples.html. @nafur what do you think?

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 27

We will ask the solver to produce models and unsat cores in the following,
and for this we have to enable the following options.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 31-32

Next we set the logic.
The simplest way to set a logic for the solver is to choose "ALL".
This enables all logics in the solver.
Alternatively, ``"QF_ALL"`` enables all logics without quantifiers.
To optimize the solver's behavior for a more specific logic,
use the logic name, e.g. ``"QF_BV"`` or ``"QF_AUFBV"``.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 41

In the following, we will define constraints of reals and integers.

> For this, we first query the solver for the corresponding sorts.

Reviewer: Is `query` the right word? sounds a bit like a check-sat query or something.

Author: It is the right word, yes. You ask the solver for the sort.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 45-46

Now, we create two constants ``x`` and ``y`` of sort ``Real``,

> and two constants ``a`` and ``b`` of sort ``Integer``.

Reviewer: What do you think about adding here a clarification about the word `constant`?
Maybe something like:
```
Notice that these are *symbolic* constants, but not actual values.
```

Author: Oh, I thought I had copied that from the example. It seems that I didn't

:cpp:func:`cvc5::api::Result::isSat`,
:cpp:func:`cvc5::api::Result::isUnsat` and
:cpp:func:`cvc5::api::Result::isSatUnknown`.
Alternatively, it can also be printed.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 96-97

This will print:

.. code:: text

  expected: sat
  result: sat

Now, we query the solver for the values for ``x`` and ``y`` that satisfy
the constraints.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 100-101

It is also possible to get values for terms that do not appear in the original
formula.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 105-106

We can retrieve the string representation of these values as follows.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 109-115

This will print the following:

.. code:: text

  value for x: 1/6
  value for y: 1/6
  value for x - y: 0

We can convert these values to C++ types using standard conversion functions.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 119-121

> Another way to independently compute the values of ``x`` and ``y`` would be

Reviewer: ```suggestion
Another way to independently compute the value of ``x-y`` would be
```

Would you mind changing this also on the cpp?

using the C++ minus operator instead of asking the solver.
However, for more complex terms, it is easier to let the solver do the
evaluation.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp

>      :lines: 119-121

Reviewer: ```suggestion
     :lines: 127-135
```

This will print:

.. code:: text

  computed correctly

Next, we will check satisfiability of the same formula,
only this time over integer variables ``a`` and ``b``.
For this, we first reset the assertions added to the solver.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 141

Next, we assert the same assertions as above, but with integers.
This time, we inline the construction of terms
to the assertion command.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 146-150

Now, we check whether the revised assertion is satisfiable.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 153

This time the asserted formula is unsatisfiable:

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp
     :lines: 156-157

> .. code:: text
> 
>   expected: unsat
>   result: unsat

Reviewer: ```suggestion
The output will be:
.. code:: text

  expected: unsat
  result: unsat
```

We can query the solver for an unsatisfiable core, that is, a subset
of the assertions that is already unsatisfiable.

.. literalinclude:: ../../../examples/api/cpp/quickstart.cpp
     :language: cpp

>      :lines: 161-166

Reviewer: ```suggestion
     :lines: 161-167
```
